### PR TITLE
8296754: AutoCreateSharedArchive in JDK 20 is not compatible with JDK 19

### PR DESCRIPTION
**Pull request by matias9927**
The -XX:+AutoCreateSharedArchive flag was implemented in JDK 19, however, this flag doesn't work across JDK 19 and 20.

Expected: JDK 20 should recreate the specified CDS archive
Actual: JDK 20 cannot recognize the archive file and gives up

The new field from GenericCDSFileMapHeader is now in FileMapHeader. Verified with tier 1-4 tests.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8296754](https://bugs.openjdk.org/browse/JDK-8296754): AutoCreateSharedArchive in JDK 20 is not compatible with JDK 19

### Reviewers
 * [Calvin Cheung](https://openjdk.org/census#ccheung) (@calvinccheung - **Reviewer**) ⚠️ Review applies to [718b4f64](https://git.openjdk.org/jdk/pull/11148/files/718b4f64c74fa613f30a5f1b5f34071c7e1839dd)
 * [Ioi Lam](https://openjdk.org/census#iklam) (@iklam - **Reviewer**)
 * [Erik Joelsson](https://openjdk.org/census#erikj) (@erikj79 - **Reviewer**) ⚠️ Review applies to [4d168999](https://git.openjdk.org/jdk/pull/11148/files/4d1689991c9c38f0b1864e2107ac13b7a98ddcfa)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/11148/head:pull/11148` \
`$ git checkout pull/11148`

Update a local copy of the PR: \
`$ git checkout pull/11148` \
`$ git pull https://git.openjdk.org/jdk pull/11148/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 11148`

View PR using the GUI difftool: \
`$ git pr show -t 11148`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/11148.diff">https://git.openjdk.org/jdk/pull/11148.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/cds/cdsConstants.cpp [hunk 1]
```diff
@@ -31,18 +31,18 @@
...
-  { "GenericCDSFileMapHeader::_crc",                       offset_of(GenericCDSFileMapHeader, _crc)            },
-  { "GenericCDSFileMapHeader::_version",                   offset_of(GenericCDSFileMapHeader, _version)        },
-  { "GenericCDSFileMapHeader::_header_size",               offset_of(GenericCDSFileMapHeader, _header_size)    },
-  { "GenericCDSFileMapHeader::_common_app_classpath_prefix_size", offset_of(GenericCDSFileMapHeader, _common_app_classpath_prefix_size) },
+  { "GenericCDSFileMapHeader::_magic",                     offset_of(GenericCDSFileMapHeader, _magic)                    },
+  { "GenericCDSFileMapHeader::_crc",                       offset_of(GenericCDSFileMapHeader, _crc)                      },
+  { "GenericCDSFileMapHeader::_version",                   offset_of(GenericCDSFileMapHeader, _version)                  },
+  { "GenericCDSFileMapHeader::_header_size",               offset_of(GenericCDSFileMapHeader, _header_size)              },
   { "GenericCDSFileMapHeader::_base_archive_name_offset",  offset_of(GenericCDSFileMapHeader, _base_archive_name_offset) },
```
Comment (calvinccheung) [inline src/hotspot/share/cds/cdsConstants.cpp:38]: I think you can delete one space between "," and "offset_of" and also adjust other lines accordingly.

File: test/hotspot/jtreg/runtime/cds/appcds/test-classes/HelloJDK8.java [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+
+// Used by tests like dynamicArchive/TestAutoCreateSharedArchive.java, which
```
Comment (calvinccheung) [inline test/hotspot/jtreg/runtime/cds/appcds/test-classes/HelloJDK8.java:26]: s/TestAutoCreateSharedArchive/TestAutoCreateSharedArchiveUpgrade

**Review by calvinccheung — APPROVED** [review]
Looks good. Just couple of nits.

File: test/hotspot/jtreg/runtime/cds/appcds/dynamicArchive/TestAutoCreateSharedArchiveUpgrade.java [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+ * @compile -source 1.8 -target 1.8 ../test-classes/HelloJDK8.java
+ * @run driver jdk.test.lib.helpers.ClassFileInstaller -jar Hello.jar HelloJDK8
+ * @run driver TestAutoCreateSharedArchiveUpgrade
+ */
+
+import java.io.File;
+import jdk.test.lib.helpers.ClassFileInstaller;
+import jdk.test.lib.process.OutputAnalyzer;
+import jdk.test.lib.process.ProcessTools;
+import jtreg.SkippedException;
```
Comment (iklam) [inline test/hotspot/jtreg/runtime/cds/appcds/dynamicArchive/TestAutoCreateSharedArchiveUpgrade.java:40]: This import is no longer necessary.

File: test/hotspot/jtreg/runtime/cds/appcds/dynamicArchive/TestAutoCreateSharedArchiveUpgrade.java [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+    }
+
+    static void assertJSAVersionMismatch(OutputAnalyzer output) {
+        output.shouldContain("does not match the required version");
+    }
+
+    static void assertUsedJSA(OutputAnalyzer output) {
+        output.shouldContain("Mapped dynamic region #0");
+    }
+}
```
Comment (iklam) [inline test/hotspot/jtreg/runtime/cds/appcds/dynamicArchive/TestAutoCreateSharedArchiveUpgrade.java:145]: Need a newline at the end of the file.

**Review by iklam — APPROVED** [review]
LGTM. Some minor nits for the test.

**Review by erikj79 — APPROVED** [review]
Makefile change looks good.

**Review by calvinccheung — APPROVED** [review]

**Review by iklam — APPROVED** [review]
LGTM

**Comment by matias9927** [comment]
Thank you for the feedback Eric, Calvin, and Ioi! 
/integrate

**Review by iklam — APPROVED** [review]
Approved latest changes again. Matias had to do second force-push which reverted the effects of the first force-push.

**Pull request closed without merging**